### ISSUE TITLE
sql: keep normalized names in table/index descriptors

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#8200]
While working on #8152 I stumbled on the following issue (Gitter transcript):

(23:07) < knz> actually I am pretty sure that table/column/index/family names are stored unnormalized in the db
(23:07) < knz> and only normalized when accessed
(23:07) < knz> this is ... strange
(23:09) < dt> that’s correct
(23:09) < dt> we’re case-preserving, case-insentive
(23:09) < knz> why case-preserving?
(23:09) < dt> same as ntfs or osx
(23:10) < dt> why not?
(23:10) < knz> it seems that the overhead to re-normalize on each query is not justified
(23:10) < knz> SQL is thoroughly case-insensitive, I do not see any benefit to store case-sensitively
(23:10) < mjibson> users like to see the same thing they typed
(23:10) < RaduBerinde> Yeah I was about to say - it looks like we don't normalize when we create, but we compare normalized name when we look up a table
(23:11) < dt> yeah, `FindColumnByName` etc all normalize as they search
(23:11) < knz> I know it _works_ but I do not understand the trade-off
(23:11) < knz> **we're doing extra work on every query for the sake of ... what exactly?**
(23:13) < RaduBerinde> Well I assume it was a purely functional decision.. I can see the value in having the names as given (e.g. in `SHOW`).
(23:14) < knz> but right now this is used inconsistently. If I do e.g. create table foo(x int, index a (X)) then I do "show create table foo" I see inconsistent "x" and "X" next to each other in the schema
(23:14) < knz> not only we're re-normalized at every query, but as you see in this specific example each insert must re-normalize all column names stored/indexed by all index
(23:15) < RaduBerinde> So is that the problem, or the runtime cost? :) (I would be surprised if the runtime cost can't be easily optimized, if indeed it's a problem)
(23:15) < knz> **I do not understand what justifies this run-time overhead**
(23:15) < RaduBerinde> I don't understand what you are saying above. On insert we have to re-normalize what?
(23:15) < knz> this seems profoundly inefficient
(23:16) < RaduBerinde> If it's profoundly inefficient, we can easly store normalized names in memory..
(23:16) < knz> **RaduBerinde: whenever an insert is running we have to find in the index descriptor which columns the index is sensitive on (and will cause an index update). This check needs to compare column names, and will re-normalize for this check.**
(23:17) < knz> What do you mean store normalized names in memory? Like re-normalize whenever we load the descriptor in memory from the db
(23:17) < knz> ?
(23:17) < RaduBerinde> Ah, yes. **We can store then normalized in memory.. In fact we could have maps from normalized name to < whatever >.**
(23:17) < RaduBerinde> Yeah (perhaps lazily)

**Comment by RaduBerinde** [comment]
f runtime cost is the only concern, we can investigate various avenues for optimization:
 - "caching" the normalized names inside descriptors
 - building maps with string keys where the string key is normalized whenever we load a descriptor in memory
 - using `strings.EqualFold(a, b)` instead of `NormalizeName(a) == NormalizeName(b)`, as in most cases `EqualFold` would exit early

**Comment by petermattis** [comment]
`strings.EqualFold(a, b)` is not equivalent to `NormalizeName(a) == NormalizeName(b)`. In particular, we normalize the turkish/azeri lowercase dotless-i and upper-case dotted-i.

**Comment by RaduBerinde** [comment]
Got it. Then - assuming case doesn't affect the UTF8 length - something along the lines of:
```
if len(str1) != len(str2) {
   return false
}
for i := 0; i < len(str1); i++ {
   a := str1[i]
   b := str2[i]
   if (ToLower(a) != ToLower(b)) {
      if (!isASCII(a) || !isASCII(b)) {
         return NormalizeName(str1[i:]) == NormalizeName(str2[i:])
      }
      return false
   }
}
```

**Comment by knz** [comment]
Fixed in #16884.
